### ISSUE TITLE
Goliath

### ISSUE THREAD

**Issue by HellDragon-HK** [GreatEmerald/UT3Vehicles#5]
Issues noticed with Goliath from in-game testing

UT2004 Draw  - 1.0?
UT3    Draw  - 1.35
UT3Port Draw - 1.0 / 1.15 but it's still not as high as UT3, standing by that red triangle on the side in UT3 is directly at head height on the port it's slightly below
  - [Priority]
  - [On UT3 the entire turret turns at the same time including the minigun, on the port only the main cannon is turning while the hatch section doesn't which isn't right and looks very odd]
  - [The minigun turret isn't staying connected to the main turret when either of them is turning]
  - [Collision is good except on the main cannon where it has none but in UT3 you can't walk through the barrel however I'm pretty sure you can with the stock UT2004 Goliath....probably more of a secondary or even tertiary issue really]
  - [Secondary]
  - [The antenna is missing on the port]
  - [The port can be pushed by the hoverboard (its a large tank so that probably shouldn't be possible)]
  - [No lights at all on the port which should be on the very front sides that cover the treads and on the front sides of the main turret (stock UT2004 Goliath also doesn't have any lights apparently although looks like it is supposed to?)]
  - [Tertiary]
  - [Port feels faster in movement than in UT3]

**Comment by GreatEmerald** [comment]
You can try solving the headlight issue here as well. You should need to set `HeadlightCoronaOffset(0)`, `HeadlightCoronaOffset(1)`, `BrakeLightOffset(0)`, `BrakeLightOffset(1)`, `bMakeBrakeLights = True` and also set the materials (perhaps use `HeadlightCoronaMaterial = Material'EpicParticles.flashflare1'` and `BrakeLightMaterial = Material'EpicParticles.flashflare1'`).

I think I matched `GroundSpeed` between the games, but it might use different units, because in UT2004 it's actually a Manta-style vehicle. Ideally it should be made into an `ONSWheeledVehicle`, but that's complicated.

The rest are missing features.

**Comment by HellDragon-HK** [comment]
I'm betting speed is actually the same, what's missing is that bounce it has when it starts moving and stops moving that sadly can't be replicated in UT2004 apparently, that would throw off comparing the speed probably.  I can try to re-compare or you can just dismiss the issue, your call.

**Comment by 100GPing100** [comment]
The reason the Goliath does not have the antenna is because I was going to implement some fake "physics" with it, just like the ailerons in the manta here:
https://github.com/GreatEmerald/UT3Vehicles/blob/master/Classes/UT3Manta.uc#L116

**Comment by HellDragon-HK** [comment]
Ok that makes sense but wasn't listed anywhere (I don't think anyway) of why it was missing so I assumed it was just unnoticed.

**Comment by 100GPing100** [comment]
Yeah, that was a fault on my side, I didn't create a `todo` list back then.

**Comment by HellDragon-HK** [comment]
@GreatEmerald 
Confirmed with my PC UT3 that Goliath does have 900 health and only 800 in UT2004

**Comment by GreatEmerald** [comment]
Try not to use the editor to raise models, that's what PrePivot is supposed to be for. Otherwise you're changing the centre of mass, which may cause complications. Not sure why that would change the collision though.

**Comment by HellDragon-HK** [comment]
I don't know where PrePivot is or how to use it

**Comment by GreatEmerald** [comment]
It's a defaultproperty, it should be possible to do `set UT3Goliath PrePivot (X=0,Y=0,Z=100)` or so. Though it might be that vehicles ignore that property like they ignore DrawScale3D...

**Comment by HellDragon-HK** [comment]
I'm guessing that has to be done in the code, it doesn't appear to work in-game.

The Scorpion and the Hellbender have the movement bounce replicated so why can't the Goliath have it, because it's a Manta type vehicle?

**Comment by GreatEmerald** [comment]
Yeap. It would need to be turned into a Paladin-style vehicle instead.

**Comment by HellDragon-HK** [comment]
Can we add that to the list, I know you said it would be complicated but it's kind of needed as the UT3 Goliath is one of the bounciest vehicles in the game in my opinion (certainly feels that way on TankCrossing when you're moving around to get a shot....and for that matter it bounces on firing too doesn't it?)

**Comment by GreatEmerald** [comment]
Sure, that's a wishlist item certainly.

**Comment by HellDragon-HK** [comment]
Added to the Extra section.  Out of curiosity if you don't mind how long before some of the issues only you and Gping can handle get implemented or are you guys already working on some of it or what's currently in the works?

**Comment by GreatEmerald** [comment]
I don't even have it installed, and probably won't have it installed until I get to LP'ing UT2004. Which by my estimation will take me around 3 years or so.
If someone else wants to work on improving it in the mean while, pull requests are welcome.

**Comment by HellDragon-HK** [comment]
Ouch, 3 years?  Well here's hoping Gping is able or someone stumbles in and joins the team like Alex has.

**Comment by HellDragon-HK** [comment]
Been trying to do the simple task of making the Goliath taller but I can't find the line of code that works, using just DrawScale = changes the whole thing instead of just making it taller like we need and trying to use X Y Z makes it invisible in-game.

PrePivot doesn't seem to have any effect in console command or in code.

We know the scale values we need from trying them in UnrealEd so how do we use those in code then, so far it doesn't seem possible to use specific values and I've found nothing of help searching the net including 3dbuzz.

**Comment by HellDragon-HK** [comment]
Well here's the 900 health fix so at least that's done now https://drive.google.com/open?id=1jDN7wWlIxAvw4sWCGFrWdmdOw6WbykWG

**Comment by GreatEmerald** [comment]
`DrawScale` is a single integer. `DrawScale3D` is a vector (three floats), but apparently that doesn't work for vehicles?
Good to hear that `PrePivot` works! That means that it should be possible to use Z translation to set the centre of mass and PrePivot to tweak the position regardless of the centre of mass.

Again, send pull requests.

**Comment by HellDragon-HK** [comment]
Either it doesn't work or I don't know how to type it correctly so far nothing works and I tried a lot of different ways of typing it in

**Comment by GreatEmerald** [comment]
Yea, it looks like the game just ignores `DrawScale3D` for vehicles specifically. In that case, feel free to just change the properties of the animation files. They will need to be sent eventually to the repository as a pull request too, but that should wait until all settings are changed so that the files don't require a second update.

**Comment by HellDragon-HK** [comment]
Changing the scale in the animation files is having serious issues when it comes down to the actual vehicle chassis, collision is getting completely removed and being replaced with a sphere, it doesn't seem possible to restore it either.  I guess this isn't happening with the turrets because they have no collision of their own?

**Comment by GreatEmerald** [comment]
That is pretty weird. Collision should be either a cylinder (like it is defined in the code) or per-polygon. Maybe there are some settings for it in UnrealEd... Else the meshes would need to be edited in some external program, and that's prone to all sorts of problems.

**Comment by HellDragon-HK** [comment]
Cylinder is a good description of what I'm experiencing, it replaces the boxes it seems with no way to get the m back which I don't understand.

**Comment by HellDragon-HK** [comment]
@GreatEmerald 
DrawScale3D does work for vehicles, I found it in the defaultproperties of ONSPRV and tried it in defaultproperties of the Goliath in the same location and changed the values to what is desired which worked.   Problem is I'm having one of the same issues as I had with editing it in Unreal Ed and it's that it will not move now but at least no collision cylinder in the way.   I also used PrePivot in-game to raise the Goliath up a bit and it still won't move, any ideas?

**Comment by HellDragon-HK** [comment]
Goliath Minigun UT3
    KDamageImpulse=700
    VehicleDamageScaling=0.5
	NodeDamageScaling=0.5

Goliath UT2004
    DamageMin=17
    DamageMax=17

How do we get around that because I'm guessing that is interferring, also the DamageScaling has not been added to the mod UT3 Goliath that I can see

**Comment by HellDragon-HK** [comment]
Added link for UT3 Drawscale ukx edit, which needs to be done in the ukx file for 2 reasons, the 1st and most important is collision actually auto-updates when the drawscale is even with all three values which is nice, the 2nd is the weapons also need the scale increase.   I also found out having the scale set in the uxk and the class at the same time doubles it instead of being the same, so any Drawscale in the class of an edited ukx needs to be // or just removed to avoid issues.

**Comment by GreatEmerald** [comment]
No, I definitely have no time to test this myself. I barely have enough time to catch up with all the issues, as you can see!

I've never seen the UT2004 Goliath with a headlight. If it existed, it must have been in one of the earliest patches.

Yes, obviously you can't change the scale in both the code and the mesh properties. I suggest keeping the scale default in mesh properties and changing it in the code, so that the numbers match UT3 numbers. Weapons should be scalable the same way. Unless there's some good reason why it would work better when scaled in the mesh.

**Comment by HellDragon-HK** [comment]
The collision boxes also get auto scaled with the mesh and don't with the code I think, doing with code I can walk through it so I'm going to say it doesn't

**Comment by HellDragon-HK** [comment]
I know you want it in code, but the ukx edits seem solid, no sinking into the floor, decent collision which may need to be larger and it doesn't really feel like it changed game play wise, although I'm not playing with bots right now and seeing them get destroyed so there might be something in that field still unknown

**Comment by HellDragon-HK** [comment]
@GreatEmerald Confirmed that code Drawscale doesn't include auto collision scaling

**Comment by GreatEmerald** [comment]
Do you mean the cylinder collision, or the surface collision? If it's just the cylinder one, you can change that in code with `CollisionHeight` and `CollisionRadius`. I think the cylinder is used when you're climbing on it, but when shooting IIRC the per-surface collision is used.

**Comment by GreatEmerald** [comment]
Oh, and the wheels are not turning because it's not a wheeled vehicle. They're supposed to but they're not going to until it gets ported to `ONSWheeledVehicle`. As for the treads getting slower, no it's not, it's just an optical illusion.

**Comment by HellDragon-HK** [comment]
No not a cylinder, the collision boxes, I haven't even seen a cylinder in use....think the boxes override that as I was getting cylinder collision when it broke trying to scale it up earlier but not at any other point that I can recall

Maybe, I'm not sure though, it looks like the treads almost completely stop with a bit of a jerk motion then slow down....it's odd, I suppose it could be an illusion but still it shouldn't jerk like that and appear t completely stop should it?

**Comment by HellDragon-HK** [comment]
@GreatEmerald Ok pieced it out into new issues, is that how you wanted it and is there anything else here that needs to be pieced out?

**Comment by GreatEmerald** [comment]
Ah yes, `DrawScale` doesn't change collision, indeed. It's quite likely that if the mesh has collision, it's used instead of the cylinder one, yes. Hm, but then doesn't that mean that in order to have correct collision, the scale needs to be changed in-editor, and `DrawScale` set to 1.0 everywhere in code?

Yes, the treads shouldn't completely stop spinning while moving.

The issue looks good now.

**Comment by HellDragon-HK** [comment]
So ukx edits are unavoidable then you're saying?

**Comment by 100GPing100** [comment]
>  Change Goliath to Paladin style vehicle so it has bounce when it moves like in UT3 or get ChassisTorque to work somehow

`ChassisTorque` does work, given the current default values, you could use something like this:
```UnrealScript
PitchTorqueFactor=-600.0;
PitchDampFactor=60.0;
```

The problem is that using this makes the Goliath uncontrollable in uneven terrain, because not all thrusters will be working at the same time, since some of them will be slightly raised.

**Comment by HellDragon-HK** [comment]
Hm that does kind of work but it's staying pitched and not just doing it on initial movement like it needs to

**Comment by 100GPing100** [comment]
@HellDragon-HK Well this kind of torque is based on the current force being applied for movement as far as I know; on wheeled vehicles what happens is the suspension getting to work.

**Comment by HellDragon-HK** [comment]
Confirmed wheels in the treads do move in UT3 and don't appear to be in UT2004

**Comment by HellDragon-HK** [comment]
Another new issue, the Goliath should take at least 1 point of damage from hitting a wall but it takes none as far as I can tell.

**Comment by HellDragon-HK** [comment]
Getting the treads sound with the engine....

I've come up with 3 approaches to this
1. Mix the sounds together and make it the engine sound which isn't ideal (what we are currently using)

2. Use the engine and engine treads mix and have them switch around based on movement, also probably not ideal because then the engine loop would suddenly change however it may not be noticeable 

3.Instead of switching the engine sounds periodically we need to add an ambient based on movement, can something like this be written into the code?

````
Simulated Function GetVelocity
{
Velocity > 0.1
PlayAmbientSound(TreadsSound, Slont_None, 1.0)
}
else
{
Velocity <= 0.1
PlayAmbientSound(None)
}
````
